azurelinuxagent/ga/agent_update.py

```diff
@@ -0,0 +1,340 @@
+
+    @staticmethod
+    def __get_available_agents_on_disk():
+        available_agents = [agent for agent in AgentUpdateHandler.__get_all_agents_on_disk() if agent.is_available]
+        return sorted(available_agents, key=lambda agent: agent.version, reverse=True)
+
+    @staticmethod
+    def __get_all_agents_on_disk():
+        path = os.path.join(conf.get_lib_dir(), "{0}-*".format(AGENT_NAME))
+        return [GuestAgent.from_installed_agent(path=agent_dir) for agent_dir in glob.iglob(path) if os.path.isdir(agent_dir)]
+
+    def __log_event(self, level, msg_, success_=True):
+        if level == LogLevel.INFO:
+            logger.info(msg_)
+            add_event(op=WALAEventOperation.AgentUpgrade, is_success=success_, message=msg_, log_event=False)
+        else:
+            msg_ += "[NOTE: Will not log the same error for the next 6 hours]"
+            # Incarnation may change if we get new goal state that would make whole string unique every time. So comparing only the substring until Incarnation if Incarnation included in msg
+            # Example msg "Unable to update Agent: No manifest links found for agent family: Prod for incarnation: incarnation_1, skipping agent update"
+            now = datetime.datetime.now()
+            prefix_msg = msg_.split("incarnation", 1)[0]
+            prefix_last_warning_msg = self.persistent_data.last_warning.split("incarnation", 1)[0]
+            if prefix_msg != prefix_last_warning_msg or self.persistent_data.last_warning_time == datetime.datetime.min or now >= self.persistent_data.last_warning_time + datetime.timedelta(hours=6):
+                if level == LogLevel.WARNING:
+                    logger.warn(msg_)
+                elif level == LogLevel.ERROR:
+                    logger.error(msg_)
+                add_event(op=WALAEventOperation.AgentUpgrade, is_success=success_, message=msg_, log_event=False)
+                self.persistent_data.last_warning_time = now
+                self.persistent_data.last_warning = msg_
+
+    def run(self, goal_state):
+        try:
+            # Ignore new agents if update is disabled. The latter flag only used in e2e tests.
+            if not self._autoupdate_enabled or not conf.get_ga_updates_enabled():
+                return
+
+            self._gs_id = goal_state.extensions_goal_state.id
+            agent_family = self.__get_agent_family_from_last_gs(goal_state)
+            requested_version = self.__get_requested_version(agent_family)
+            agent_manifest = None  # This is to make sure fetch agent manifest once per update
+
+            if requested_version is None:
+                if conf.get_enable_ga_versioning():  # log the warning only when ga versioning is enabled
+                    warn_msg = "Missing requested version in agent family: {0} for incarnation: {1}, fallback to largest version update".format(self._ga_family, self._gs_id)
+                    self.__log_event(LogLevel.WARNING, warn_msg)
```

Comment:
i don't think this would swamp the log, if goal states were that frequent then we have bigger problems, since we log quite a few messages in each goal state.

now, coming to think about it, we should probably log it when we actually attempt the update, otherwise i think the message is not interesting